examples/Example3/physics.cc

```diff
@@ -0,0 +1,118 @@
+// SPDX-FileCopyrightText: 2020 CERN
+// SPDX-License-Identifier: Apache-2.0
+
+#include "physics.h"
+
+#include "particle.h"
+#include "point.h"
+#include "random.h"
+#include "vector.h"
+
+#include <cmath>
+#include <iostream>
+
+using std::exp;
+using std::log;
+using std::sqrt;
+
+#define Ecut 10.0f
+
+__host__ __device__ float absorption(float E)
+{
+	return (5000.0f / E) * exp(-0.001 * E / 2.0);
+}
+
+__host__ __device__ float scattering(float E)
+{
+	return E <= 100.0f ? 0.0f : 10.0f * (0.001 * E - 0.1f) * exp(-sqrt(0.001 * E));
+}
+
+__host__ __device__ float production(float E)
+{
+	return E < 100.0f ? 0.0f : 0.01 * std::log(E - 9.0f);
+}
+
+__host__ __device__ float cross_section(float E)
+{
+	return absorption(E) + scattering(E) + production(E);
+}
+
+__host__ __device__ float sample_interaction_length(particle &p)
+{
+	return -log(uniform()) / cross_section(p.energy());
+}
+
+__host__ __device__ void energy_loss(particle &p)
+{
+	if (std::abs(p.charge) > 0.0f)
+		p.velocity *= 0.99;
+}
+
+__host__ __device__ void absorption_interaction(particle &p)
+{
+	p.alive = false;
+	debug_printf("# absorbed\n");
+}
+
+__host__ __device__ void scattering_interaction(particle &p)
+{
+	float x = uniform(-1.0f, 1.0f);
+	float y = uniform(-1.0f, 1.0f);
+	float z = uniform(-1.0f, 1.0f);
+	vector dv{x, y, z};
+	dv = cross(p.velocity, dv / norm(p.velocity));
+	vector v = (p.velocity + dv) / norm(p.velocity + dv);
+	p.velocity = norm(p.velocity) * v;
+	debug_printf("# scattered\n");
+}
+
+__host__ __device__ void production_interaction(particle &p)
+{
+	debug_printf("# pair production\n");
+	particle p1(p);
+	particle p2(p);
+	p1.time = 0.0f;
+	p2.time = 0.0f;
+
+	float n = norm(p.velocity);
+	float s = uniform(0.1f, 0.9f);
+	vector v = p.velocity / norm(p.velocity);
+	vector u = random_unit_vector();
+	p1.velocity = 0.5 * n * (v + s * u);
+	p2.velocity = 0.5 * n * (v - (1.0f - s) * u);
+
+	if (p.charge == 0.0f) {
+		p1.charge = +1.0f;
+		p2.charge = -1.0f;
+	} else {
+		p1.charge = 0.0f;
+		p2.charge = 0.0f;
+	}
+
+	create_particle(p1);
+	create_particle(p2);
+	p.alive = false;
+}
+
+__host__ __device__ void interact(particle &p)
+{
+	float E = p.energy();
+
+	if (E < Ecut) {
+		debug_printf("# energy is below minimum\n");
+		p.alive = false;
+	}
+
+	float inv_crosssec = 1.0f / cross_section(E);
+	float p_absorption = absorption(E) * inv_crosssec;
+	float p_scattering = scattering(E) * inv_crosssec;
+
+	float r = uniform();
+
+	if (r < p_absorption)
```

Comment:
My understanding is that G4HepEM approach is also a bit like this. I mean for each particle type, there is a the process manager which has the 'Perform' method which calls hardwired sequence of processes. So we really need to understand what we want to be a kernel: a sequence of processes for a given particle type or each process separately? Or somehow a combination of the two?

Reply:
If all processes look the same for a given particle in Mihaly's approach this is indeed the case. However, the Perform method does different things depending on the selected process  (e.g. generating and handling secondaries). Also, don't forget transportation that we don't want to be mixed with physics. I mean, anything is possible, but the optimum efficiency may be a compromise in between.

Reply:
Right now the whole step is quite small, and therefore kept inside a single kernel call. As the code evolves, it will be absolutely necessary to split the stepping into multiple parts/kernels, of course. For physics processes, I think that one kernel per particle type for computing the interaction length should be good enough, while for processing secondaries, a more complex approach may be needed (i.e. further splitting into more kernels), but we won't know that until we have enough code to support complex physics (including materials, etc), so for now we only need to make sure that we design the code in a way that this particular aspect can be easily changed in the future.